networkx/algorithms/approximation/steinertree.py

```diff
@@ -46,22 +57,172 @@ def metric_closure(G, weight="weight"):
     return M
 
 
+def mehlhorn_steiner_tree(G, S, weight):
+    paths = nx.multi_source_dijkstra_path(G, S)
+
+    d_1 = {}
+    s = {}
+    for v in G.nodes():
+        s[v] = paths[v][0]
+        d_1[(v, s[v])] = len(paths[v]) - 1
+
+    new_edges = {}
+    for e in G.edges(data=True):
+        u, v, data = e
+        weight_here = d_1[(u, s[u])] + data.get(weight, 1) + d_1[(v, s[v])]
+        key = tuple(sorted([s[u], s[v]]))
+        new_edges[key] = min(new_edges.get(key, float("inf")), weight_here)
+
+    G_1_prime = nx.Graph()
+    for (n1, n2), edge_weight in new_edges.items():
+        G_1_prime.add_edge(n1, n2, weight=edge_weight)
+
+    G_2 = nx.minimum_spanning_edges(G_1_prime, data=True)
+
+    G_3 = nx.Graph()
+    for u, v, d in G_2:
+        path = nx.shortest_path(G, u, v, weight)
+        for n1, n2 in pairwise(path):
+            G_3.add_edge(n1, n2)
```

Comment:
Can G_1_prime be called G_1?  not a big deal, but it is a little strange.

Should G_2 be called G at all?  It is a bunch of edges... not a graph.
Either way, you don't seem to use "data=True" set in the "d" variable in the next loop, but then not used.
So, maybe something like:

```python
    G_3 = nx.Graph()
    for u, v in nx.minimum_spanning_edges(G_1_prime, data=False):
        G_3.add_edges_from(pairwise(nx.shortest_path(G, u, v, weight)))

    return nx.minimum_spanning_edges(G_3, data=False)
```

Reply:
Ahh... I see.
Maybe add a comment at line 75 like: 
`# G1-G4 names match those from the Mehlhorn 1988 paper.`

It's a little strange to name a list of edges G1_prime as if it is a graph.  Does the paper also have G1? what about another G with a prime? What does the prime signify?

Reply:
Kou and Mehlhorn's algorithms are basically the same, but with a different first step of the total five steps. This G1_prime of Mehlhorn is the replacement of Kou's G1.

With this in mind, I think the methods for the two of them could actually be merged, but maybe that's a matter for another PR?

Reply:
Sounds good to postpone any decisions about merging the two methods.